### PR TITLE
Get username and password from .netrc file if available

### PR DESCRIPTION
**Pull request by adybbroe**
Check if username and password are available in the .netrc file for the host of the destination

### CODE REVIEW

File: trollmoves/movers.py [hunk 1]
```diff
@@ -42,6 +43,12 @@
 
 LOGGER = logging.getLogger(__name__)
 
+HOME = os.environ.get('HOME')
+if HOME:
+    NETRCFILE = os.path.join(HOME, '.netrc')
+else:
+    NETRCFILE = None
+
```
Comment (mraspaud) [inline trollmoves/movers.py:51]: This isn't needed. According to the documentation of netrc, the filename is optional and defaults to home's .netrc: https://docs.python.org/3.8/library/netrc.html#netrc.netrc

File: trollmoves/movers.py [hunk 2]
```diff
@@ -201,10 +209,30 @@ class FtpMover(Mover):
...
+                    self.destination.hostname)
+                logger.debug('Got username and password from netrc file!')
+
     def open_connection(self):
+        """Open the connection and login."""
         connection = FTP(timeout=10)
         connection.connect(self.destination.hostname,
                            self.destination.port or 21)
+
+        self._get_netrc_authentication()
```
Comment (mraspaud) [inline trollmoves/movers.py:235]: This should be moved a couple of lines further down, so as not to overwrite username and password if they were provided in `destination`

File: trollmoves/movers.py (hunk 1 above)
Comment (adybbroe) [reply trollmoves/movers.py:51]: Well, you are right. However, this came from experience with another application where HOME is not defined. As you know when run from supervisor HOME is not available. Therefore I implemented this. If home is not set you will get the following in your case:
```
In [5]: secrets = netrc.netrc()                                                                                                                                               
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
<ipython-input-5-211a7aaf8c74> in <module>
----> 1 secrets = netrc.netrc()

~home/a000680/miniconda3/envs/python38/lib/python3.8/netrc.py in __init__(self, file)
     27         self.hosts = {}
     28         self.macros = {}
---> 29         with open(file) as fp:
     30             self._parse(file, fp, default_netrc)
     31 

FileNotFoundError: [Errno 2] No such file or directory: '/.netrc'
```
@mraspaud Do you suggest I catch that exception then and log a warning?

Comment (mraspaud) [reply trollmoves/movers.py:51]: if `$HOME` isn't available, the `netrc` module will surely raise an exception, why not just catch it ?

Comment (mraspaud) [reply trollmoves/movers.py:51]: So the answer to your question is 'yes'.

Comment (mraspaud) [reply trollmoves/movers.py:51]: also note that is the `.netrc` file isn't present in `$HOME`, you'll have to catch a `FileNotFoundError`

File: trollmoves/movers.py (hunk 2 above)
Comment (adybbroe) [reply trollmoves/movers.py:235]: Ok. Will do. Please note that _get_netrc_authentication only change the username and password if the netrc file is available and the hostname is included there.

File: trollmoves/movers.py (hunk 1 above)
Comment (adybbroe) [reply trollmoves/movers.py]: @mraspaud Can you please give me a hint where you would have such a small test? I find no tests for movers.py, so I suppose it somewhere in test_dispatcher.py? These tests seem to test on a higher level, or?

Comment (mraspaud) [reply trollmoves/movers.py]: I didn't realise there wasn't any test for movers. Could you create a test file for movers and place just a simple test there for the ftpmover's change you made ? If the file is there, we can then populate it with more tests

**Comment by adybbroe** [comment]
I pushed something accidentally. I was testing furl, as with urlparse I can't change username or password:
```
In [1]: from six.moves.urllib.parse import urlparse                                                                                                                           
In [2]: destination = 'ftp://lxserv1096.smhi.se/san1/test_msb_dispatch/afimg_20200604_125613.geojson'                                                                                                                                                                 
In [3]: dest_url = urlparse(destination)
In [4]: dest_url.username                                                                                                                                 
In [5]: dest_url.username = 'xxx'                                                                                                                                             
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-588222ecef73> in <module>
----> 1 dest_url.username = 'xxx'

AttributeError: can't set attribute
```

File: trollmoves/tests/test_movers.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    username = 'myusername'
+    password = 'mypasswd'
+    account = None
+
+    origin = '/path/to/mydata/filename.ext'
+
+    def raise_my_error():
+        raise FileNotFoundError
+
+    netrc.side_effect = raise_my_error
```
Comment (pnuu) [inline trollmoves/tests/test_movers.py:43]: The side effect can be directly set to `FileNotFoundError`, no need to wrap it in a function.

File: trollmoves/movers.py [hunk 4]
```diff
@@ -201,13 +211,38 @@ class FtpMover(Mover):
...
+        except (netrc.NetrcParseError, FileNotFoundError) as e__:
+            LOGGER.warning('Failed retrieve authentification details from netrc file! Exception: ', str(e__))
+            return
+
+        LOGGER.debug("hosts: %s", str(list(secrets.hosts.keys())))
+        LOGGER.debug("Check if hostname matches any listed in the netrc file")
+        if self.destination.hostname in list(secrets.hosts.keys()):
+            LOGGER.debug("Destination hostname: %s", self.destination.hostname)
+            self._dest_username, account, self._dest_password = secrets.authenticators(self.destination.hostname)
+            LOGGER.debug("Destination hostname: %s", self.destination.hostname)
```
Comment (pnuu) [inline trollmoves/movers.py:228]: This log message is done twice.

File: trollmoves/movers.py [hunk 5]
```diff
@@ -201,13 +211,38 @@ class FtpMover(Mover):
...
+    def _get_netrc_authentication(self):
+        """Get login authentications from netrc file if available"""
+        LOGGER.debug("HOME = %s", os.environ.get('HOME'))
+        try:
+            secrets = netrc.netrc()
+        except (netrc.NetrcParseError, FileNotFoundError) as e__:
+            LOGGER.warning('Failed retrieve authentification details from netrc file! Exception: ', str(e__))
+            return
+
+        LOGGER.debug("hosts: %s", str(list(secrets.hosts.keys())))
```
Comment (pnuu) [inline trollmoves/movers.py:223]: Is this log message necessary?

File: trollmoves/movers.py [hunk 6]
```diff
@@ -201,13 +211,38 @@ class FtpMover(Mover):
     active_connections = dict()
     active_connection_lock = Lock()
 
+    def _get_netrc_authentication(self):
+        """Get login authentications from netrc file if available"""
+        LOGGER.debug("HOME = %s", os.environ.get('HOME'))
```
Comment (pnuu) [inline trollmoves/movers.py:216]: Is this log message necessary?

File: trollmoves/movers.py (hunk 4 above)
Comment (adybbroe) [reply trollmoves/movers.py:228]: Oh, yes, it is probably quite verbose now, as I needed some indications on where it crached until I found the error... I will reduce!

File: trollmoves/movers.py (hunk 5 above)
Comment (adybbroe) [reply trollmoves/movers.py:223]: Well, hopefully nut, I just needed extra debugging info during development

File: trollmoves/tests/test_movers.py (hunk 3 above)
Comment (adybbroe) [reply trollmoves/tests/test_movers.py:43]: Hmm, this was the way I got it working, spent some time on this already :-(
This doesn't work:
```
E       netrc.side_effect = raise FileNotFoundError
E                           ^
E   SyntaxError: invalid syntax
```
How do you mean @pnuu ?

Comment (mraspaud) [reply trollmoves/tests/test_movers.py:43]: just do
```
netrc.side_effect = FileNotFoundError
```
should work in principle

**Pull request merged**